oauth.go

```diff
@@ -56,54 +55,85 @@ type OAuthDatastoreProvider interface {
+	ValidateOAuthState(context.Context, string) (string, string, error)
+	GenerateOAuthState(context.Context, string, string) (string, error)
+
 	CreateUser(*config.Config, *User, string) error
-	RecordRemoteUserID(context.Context, int64, int64) error
 	GetUserForAuthByID(int64) (*User, error)
 }
 
 type HttpClient interface {
 	Do(req *http.Request) (*http.Response, error)
 }
 
+type oauthClient interface {
+	GetProvider() string
+	GetClientID() string
+	buildLoginURL(state string) (string, error)
+	exchangeOauthCode(ctx context.Context, code string) (*TokenResponse, error)
+	inspectOauthAccessToken(ctx context.Context, accessToken string) (*InspectResponse, error)
+}
+
 type oauthHandler struct {
-	Config     *config.Config
-	DB         OAuthDatastore
-	Store      sessions.Store
-	HttpClient HttpClient
+	Config      *config.Config
+	DB          OAuthDatastore
+	Store       sessions.Store
+	oauthClient oauthClient
 }
 
-// buildAuthURL returns a URL used to initiate authentication.
-func buildAuthURL(db OAuthDatastore, ctx context.Context, clientID, authLocation, callbackURL string) (string, error) {
-	state, err := db.GenerateOAuthState(ctx)
+func (h oauthHandler) viewOauthInit(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	state, err := h.DB.GenerateOAuthState(ctx, h.oauthClient.GetProvider(), h.oauthClient.GetClientID())
 	if err != nil {
-		return "", err
+		failOAuthRequest(w, http.StatusInternalServerError, "could not prepare oauth redirect url")
 	}
-
-	u, err := url.Parse(authLocation)
+	location, err := h.oauthClient.buildLoginURL(state)
 	if err != nil {
-		return "", err
+		failOAuthRequest(w, http.StatusInternalServerError, "could not prepare oauth redirect url")
+		return
+	}
+	http.Redirect(w, r, location, http.StatusTemporaryRedirect)
+}
+
+func configureSlackOauth(r *mux.Router, app *App) {
+	if app.Config().SlackOauth.ClientID != "" {
+		oauthClient := slackOauthClient{
+			ClientID:         app.Config().SlackOauth.ClientID,
+			ClientSecret:     app.Config().SlackOauth.ClientSecret,
+			TeamID:           app.Config().SlackOauth.TeamID,
+			CallbackLocation: app.Config().App.Host + "/oauth/callback",
+			HttpClient:       &http.Client{Timeout: 10 * time.Second},
+		}
+		configureOauthRoutes(r, app, oauthClient)
 	}
-	q := u.Query()
-	q.Set("client_id", clientID)
-	q.Set("redirect_uri", callbackURL)
-	q.Set("response_type", "code")
-	q.Set("state", state)
-	u.RawQuery = q.Encode()
+}
 
-	return u.String(), nil
+func configureWriteAsOauth(r *mux.Router, app *App) {
+	if app.Config().WriteAsOauth.ClientID != "" {
+		oauthClient := writeAsOauthClient{
+			ClientID:         app.Config().WriteAsOauth.ClientID,
+			ClientSecret:     app.Config().WriteAsOauth.ClientSecret,
+			ExchangeLocation: app.Config().WriteAsOauth.TokenLocation,
+			InspectLocation:  app.Config().WriteAsOauth.InspectLocation,
+			AuthLocation:     app.Config().WriteAsOauth.AuthLocation,
```

Comment:
I think we should fall back to the default Write.as endpoints here if config values are empty. Thoughts?

Reply:
Yep, good call.